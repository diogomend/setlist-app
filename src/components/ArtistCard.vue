<template>
  <v-card
    class="mx-auto mt-5"
    max-width="450"
    dark
    :to="{ name: 'Artist', params: { mbid } }"
  >
    <v-img class="white--text align-end" height="250px" :src="getImage()">
      <v-card-title color="deep-orange lighten-1">{{ name }}</v-card-title>
    </v-img>
    <v-card-text class="text--primary">
      <v-layout row wrap class="text-left">
        <v-flex class="align-self-center ml-5">
          <div class="top-bands-attended">
            Attended <b>{{ totalShows }}</b> times
          </div>
        </v-flex>
      </v-layout>
    </v-card-text>
  </v-card>
</template>
<script>
import { mapGetters } from "vuex";
export default {
  name: "ArtistCard",
  props: {
    name: {
      type: String,
      required: true
    },
    mbid: {
      type: String,
      required: true
    },
    color: {
      type: String,
      required: false,
      default: "#967355"
    }
  },
  computed: {
    ...mapGetters({
      artistShows: "setlist/getArtistShows"
    }),
    totalShows() {
      return this.artistShows(this.mbid).length;
    }
  },
  methods: {
    getImage() {
      return `${process.env.VUE_APP_API}fanart/${this.mbid}`;
    }
  }
};
</script>

<style scoped lang="scss">
/deep/ .v-responsive__sizer {
  min-height: 100%;
  background: rgba(0, 0, 0, 0.2);
}

/deep/ .v-card__title {
  background: linear-gradient(
    to top,
    rgba(30, 30, 30, 1) 30%,
    rgba(0, 0, 0, 0)
  );
  font-size: 23px;
  word-break: break-word;
}

a.theme--dark.v-card.v-card {
  margin: 0;

  @media screen and (min-width: 600px) and (max-width: 959px) {
    margin: 24px !important;
  }
}

/deep/ .v-image__image--cover {
  background-position: top center !important;
}
</style>
