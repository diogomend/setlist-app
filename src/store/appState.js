const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  }
};

const mutations = {
  TOGGLE_DRAWER(state, bool) {
    state.drawer = bool;
  },
  TOGGLE_LOADING(state, isLoading) {
    state.isLoading = isLoading;
  }
};

const getters = {
  getDrawer: state => state.drawer,
  getLoading: state => state.isLoading
};

const state = () => ({
  drawer: false,
  isLoading: true
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
