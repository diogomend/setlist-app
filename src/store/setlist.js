import setlistService from "@/service/setlist";

const actions = {
  fetchSetlists: async ({ commit, dispatch }) => {
    try {
      const { data } = await setlistService.getSetlists();
      commit("UPDATE_SETLISTS", data);
      commit("UPDATE_ARTISTS", data);
      //return Promise.resolve(data);
      return true;
    } catch (err) {
      //return Promise.reject('Could not load setlist information');
      return false;
    } finally {
      dispatch("appState/toggleLoading", false, { root: true })
    }
  }
};

const mutations = {
  UPDATE_SETLISTS(state, data) {
    state.setlists = data;
  },
  UPDATE_ARTISTS(state, data) {
    state.artists = getArtists(data);
  }
};

const getters = {
  getTotalConcerts: state => state.setlists.length,
  getShow: state => id => {
    return Object.values(state.setlists).find(element => element.id === id);
  },
  getArtist: state => mbid => {
    const data = Object.values(state.setlists).find(
      element => element.artist.mbid === mbid
    );
    if (data && Object.prototype.hasOwnProperty.call(data, "artist")) {
      return data.artist;
    }

    return null;
  },
  getArtists: state => {
    const artists = getArtists(state.setlists);

    return Object.values(artists).sort((a, b) => (a.name > b.name ? 1 : -1));
  },
  getArtistShows: state => mbid => {
    return state.setlists.reduce((ac, setlist) => {
      if (setlist.artist.mbid === mbid) {
        ac.push(setlist);
      }

      return ac;
    }, []);
  },
  getTopArtists: state => {
    const artists = Object.values(state.artists).sort((a, b) =>
      a.qty < b.qty ? 1 : -1
    );
    return artists.slice(0, 6);
  },
  getTotalArtists: state => Object.keys(state.artists).length,
  getYearlyConcerts: state => {
    return state.setlists.reduce((ac, setlist) => {
      const dateSplit = setlist.eventDate.split("-");
      const year = parseInt(dateSplit[2]);

      if (Object.prototype.hasOwnProperty.call(ac, year)) {
        const merge = [...ac[year], setlist];
        return { ...ac, [year]: merge };
      }

      return { ...ac, [year]: [setlist] };
    }, {});
  },
  getTotalCountries: state => {
    const countries = state.setlists.reduce((ac, setlist) => {
      const country = setlist.venue.city.country.code;
      const qty = Object.prototype.hasOwnProperty.call(ac, country)
        ? ac[country] + 1
        : 1;

      return { ...ac, [country]: qty };
    }, {});

    return Object.keys(countries).length;
  }
};

const state = () => ({
  setlists: [],
  isLoading: true
});

const getArtists = setlists => {
  return setlists.reduce((ac, setlist) => {
    setlist.artist.qty = Object.prototype.hasOwnProperty.call(
      ac,
      setlist.artist.mbid
    )
      ? ac[setlist.artist.mbid].qty + 1
      : 1;

    return { ...ac, [setlist.artist.mbid]: setlist.artist };
  }, {});
};

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
