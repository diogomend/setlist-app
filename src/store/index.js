import Vue from "vue";
import Vuex from "vuex";
import setlist from "@/store/setlist";
import appState from "@/store/appState";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    setlist,
    appState
  }
};

export default new Vuex.Store(storeOptions);
