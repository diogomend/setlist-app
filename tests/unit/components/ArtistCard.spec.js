import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import { storeOptions } from "@/store/index";
import ArtistCard from "@/components/ArtistCard";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuex);
Vue.use(Vuetify);
describe("ArtistCard.vue", () => {
  let store;
  let localVue;

  beforeEach(() => {
    jest.clearAllMocks();
    store = new Vuex.Store(storeOptions);
  });

  it("is a Vue instance", () => {
    const wrapper = shallowMount(ArtistCard, {
      localVue,
      store,
      propsData: {
        name: "MOCK_NAME",
        mbid: "MOCK_MBID"
      }
    });

    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });
});
