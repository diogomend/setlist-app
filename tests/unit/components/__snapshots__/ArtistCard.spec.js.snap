// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistCard.vue is a Vue instance 1`] = `
<v-card-stub loaderheight="4" to="[object Object]" tag="div" maxwidth="450" dark="true" class="mx-auto mt-5">
  <v-img-stub height="250px" options="[object Object]" position="center center" src="http://localhost:8000/fanart/MOCK_MBID" transition="fade-transition" class="white--text align-end">
    <v-card-title-stub color="deep-orange lighten-1">MOCK_NAME</v-card-title-stub>
  </v-img-stub>
  <v-card-text-stub class="text--primary">
    <v-layout-stub tag="div" row="" wrap="" class="text-left">
      <v-flex-stub tag="div" class="align-self-center ml-5">
        <div class="top-bands-attended">
          Attended <b>0</b> times
        </div>
      </v-flex-stub>
    </v-layout-stub>
  </v-card-text-stub>
</v-card-stub>
`;
